cmd/ebpf/xdp/n3n6_entrypoint.c

```diff
@@ -40,33 +41,142 @@
 #include "xdp/utils/icmp.h"
 
 
-
 #define DEFAULT_XDP_ACTION XDP_PASS
 
+
 static __always_inline enum xdp_action send_to_gtp_tunnel(struct packet_context *ctx, int srcip, int dstip, __u8 tos, int teid) {
     if (-1 == add_gtp_over_ip4_headers(ctx, srcip, dstip, tos, teid))
         return XDP_ABORTED;
-  
     upf_printk("upf: send gtp pdu %pI4 -> %pI4", &ctx->ip4->saddr, &ctx->ip4->daddr);
     increment_counter(ctx->n3_n6_counter, tx_n3);
     return route_ipv4(ctx->xdp_ctx, ctx->eth, ctx->ip4);
 }
 
-static __always_inline enum xdp_action handle_n6_packet_ipv4(struct packet_context *ctx) {
+static __always_inline __u8 match_sdf_filter_ipv4(struct packet_context *ctx, struct sdf_filter *sdf) {
+    const struct iphdr *outer_ip4 = ctx->ip4; // Preserve the outer IPv4 header
+    int l4_protocol = parse_ip4(ctx);         // Parse the inner IPv4 header
```

Comment:
Still the same issue here. `parse_ip4` overrides ctx->ip4 value with inner ip4. Later (during routing, for instance) the program accesses `ctx->ip4`, but gets unexpected result.
You need to parse inner header in a such way that doesn't affect `ctx` at all.